batchflow/models/torch/base.py

```diff
@@ -916,84 +891,82 @@ def train(self, *args, feed_dict=None, fetches=None, use_lock=True, train_mode='
+        self.iter_info.update({
+            'microbatch': microbatch,
+            'sync_frequency': sync_frequency,
+            'steps': steps,
+            'actual_model_inputs_shape': [get_shape(item) for item in _inputs],
+            'actual_model_outputs_shape': get_shape(_targets),
+        })
         return output
 
-    def _train(self, *args, fetches=None, train_mode='', accumulate_grads=True, sync_frequency=True):
+    def _train(self, *args, fetches=None, sync_frequency=True):
+        # Parse inputs
         *inputs, targets = args
         inputs = inputs[0] if isinstance(inputs, (tuple, list)) and len(inputs) == 1 else inputs
 
-        output_container = {}
-
-        if not accumulate_grads:
-            predictions = self.model(inputs)
-
-        for mode in train_mode:
-            if mode in self.train_steps.keys():
-                train_fetches = [(mode, self.train_steps[mode])]
-            else:
-                train_fetches = [(name, train_step) for name, train_step in self.train_steps.items()
-                                 if re.search(mode, name) is not None]
-            mode_loss = 0
-
-            for name, step in train_fetches:
-                loss_fn, optimizer = step['loss'], step['optimizer']
-                decays, decay_steps = step['decay'], step['decay_step']
-                if 'initialized' not in step:
-                    optimizer.zero_grad()
-                    step['initialized'] = True
-                if accumulate_grads:
-                    predictions = self.model(inputs)
-                loss = sum([loss_fn_(predictions, targets) for loss_fn_ in loss_fn]) / len(loss_fn)
-                mode_loss += loss
-                loss.backward()
-                step['iter'] = step.get('iter', 0.0) + (1 / sync_frequency)
-
-                if self.sync_counter >= sync_frequency:
-                    for p in self.model.parameters():
-                        if p.grad is not None:
-                            p.grad /= sync_frequency
-
-                    optimizer.step()
-                    optimizer.zero_grad()
-                    self.sync_counter = 1
-                else:
-                    self.sync_counter += 1
-
-                curr_lr = [group['lr'] for group in optimizer.param_groups]
-                self.iter_info.setdefault('lr', []).append(curr_lr)
-
-                if decays:
-                    for decay, decay_step in zip(decays, decay_steps):
-                        step_condition = int(1 + step['iter'] - decay_step['first_iter']) % decay_step['frequency'] == 0
-                        range_condition = decay_step['first_iter'] <= step['iter'] <= decay_step['last_iter']
-                        if step_condition and range_condition:
-                            decay.step()
+        # Apply model, compute loss and gradients
+        predictions = self.model(inputs)
+        loss = sum(loss_fn(predictions, targets) for loss_fn in self.loss) / len(self.loss)
```

Comment:
same expression is used in `predict` - whdt about extracting a method?

Reply:
I thought about splitting the procedure of training into multiple blocks so you could configure each of them in your inherited class:

- `compute_predictions`

- `compute_loss`

- `optimizer_step`

etc. I suggest that we do it in separate PR, as it would make the diff in this one unreadable